scipy/constants/__init__.py

```diff
@@ -319,21 +319,22 @@
 # Modules contributed by BasSw (wegwerp@gmail.com)
 from ._codata import *
 from ._constants import *
-from ._codata import _obsolete_constants
+from ._codata import _obsolete_constants, physical_constants
 
 # Deprecated namespaces, to be removed in v2.0.0
 from . import codata, constants
 
-_constant_names = [(_k.lower(), _k, _v)
-                   for _k, _v in physical_constants.items()
-                   if _k not in _obsolete_constants]
+_constant_names_list = [(_k.lower(), _k, _v)
+                        for _k, _v in physical_constants.items()
+                        if _k not in _obsolete_constants]
 _constant_names = "\n".join(["``%s``%s  %s %s" % (_x[1], " "*(66-len(_x[1])),
                                                   _x[2][0], _x[2][1])
-                             for _x in sorted(_constant_names)])
+                             for _x in sorted(_constant_names_list)])
 if __doc__:
     __doc__ = __doc__ % dict(constant_names=_constant_names)
 
 del _constant_names
```

Comment:
I suppose this should be removed then.
```suggestion
```

Reply:
Not quite, the temporary `_constant_names` string still exists, it' just that the string and list-of-tuples now have distinct variable names. 

Another option would to just create a small helper function to handle the `__doc__`-formatting, which honestly doesn't like a terrible idea all things considered.

Reply:
Oh yes L330, got confused with the diff here 🤦 It's fine like this, for now I guess. I would leave further rewrite for another PR as right now it's pretty well contained with type hints changes. BTW, I supposed that this change was necessary for mypy to know what was happening right?

Reply:
> BTW, I supposed that this change was necessary for mypy to know what was happening right?

Yes, as a general of thumb type checkers don't like reusing variable-names for objects with different types. Depending on the situation you could get away with it if the type is explicitly declared as a union, but this quickly becomes difficult if the the union-members are vastly different types (like in this of this `_constant_names` variable).

``` python
from __future__ import annotations

a: list[str] | str = ["test"]  # Ok
a = "test"  # Ok
a = 1  # Not ok
```

Reply:
Ok thanks for confirming! And the details.